taskcluster/scripts/pipeline/train_taskcluster.py

```diff
@@ -1,14 +1,101 @@
 #!/usr/bin/env python3
 
+import logging
+import os
 import os.path
+import requests
 import subprocess
 import sys
 
 TRAINING_SCRIPT = os.path.join(os.path.dirname(__file__), "train-taskcluster.sh")
+CONTINUATION_ARTIFACTS = {
+    "config.opustrainer.yml",
+    "config.opustrainer.yml.state",
+    "devset.out",
+    "model.npz",
+    "model.npz.best-bleu-detok.npz",
+    "model.npz.best-bleu-detok.npz.decoder.yml",
+    "model.npz.best-ce-mean-words.npz",
+    "model.npz.best-ce-mean-words.npz.decoder.yml",
+    "model.npz.best-chrf.npz",
+    "model.npz.best-chrf.npz.decoder.yml",
+    "model.npz.decoder.yml",
+    "model.npz.optimizer.npz",
+    "model.npz.progress.yml",
+    "model.npz.yml",
+    "opustrainer.log",
+    "train.log",
+    "valid.log",
+    "vocab.spm",
+}
 
 
 def main(args):
-    subprocess.run([TRAINING_SCRIPT, *args], check=True)
+    logging.basicConfig(level=logging.INFO)
+
+    script_args = list(args)
+    task_id = os.environ["TASK_ID"]
+    run_id = int(os.environ["RUN_ID"])
+    root_url = os.environ["TASKCLUSTER_ROOT_URL"]
+    # Must line up with where model_dir is in `train-taskcluster.sh` while that script
+    # still exists.
+    model_dir = script_args[6]
+    pretrained_model_mode = None
+    if len(args) >= 11:
+        pretrained_model_mode = script_args[10]
+
+    if not os.path.exists(model_dir):
+        os.makedirs(model_dir)
+
+    if run_id > 0:
+        logging.info("run_id > 0, attempting to resume training from an earlier run...")
+        prev_run_id = run_id - 1
+
+        while prev_run_id >= 0:
+            resp = requests.get(
+                f"{root_url}/api/queue/v1/task/{task_id}/runs/{prev_run_id}/artifacts"
+            )
+            resp.raise_for_status()
+
+            run_artifacts = set([os.path.basename(a["name"]) for a in resp.json()["artifacts"]])
+
+            if run_artifacts.issuperset(CONTINUATION_ARTIFACTS):
+                logging.info(
+                    f"Run {prev_run_id} appears to have the artifacts we need! Downloading them..."
+                )
+            else:
+                logging.info(f"Run {prev_run_id} is missing some necessary artifacts...")
+                prev_run_id -= 1
+                continue
+
+            for artifact in resp.json()["artifacts"]:
+                # Skip Taskcluster logs - we only care about artifacts that the training tools create.
+                if artifact["name"].startswith("public/log"):
+                    continue
+                out_name = os.path.basename(artifact["name"])
+                logging.info(f"Fetching {artifact['name']}...")
+
+                r = requests.get(
+                    f"{root_url}/api/queue/v1/task/{task_id}/runs/{prev_run_id}/artifacts/{artifact['name']}",
+                    stream=True,
+                )
+                r.raise_for_status()
```

Comment:
I wonder what happens if there's an exception in this code, say from data downloading. Will the task just be retried? Do we want to implement any retries right here?